### ISSUE TITLE
huge memory leak

### ISSUE THREAD

**Issue by bmorel** [ValyriaTear/ValyriaTear#463]
Ok, while finally playing, I have noticed a huge memory leak. In less than 3 hours, VT takes more than 3Gb of RAM. Almost 4.
I have restarted the game, keeping an eye on the moments my free RAM takes a kick in... ok, where it does not like to, and my little finger said me that it might be when fights occur. 
Searching for more info in the source code, I have noticed that there is lot of code duplication concerning memory, which could be avoided. 
It may not fix the leak, but I would like to fix this copy/copy paste code, which could help. Those duplications are located around the calls to malloc/realloc/new, most of them in engine/video. They concern image loading, and considering that graphical resources are actually what's the heavier in games...

About that: last time I asked, I think C++11 was not allowed. Since things were moved to SDL2 in the meantime, which is more recent (I'll have to keep my backported sdl2 stuff, or to move to Debian Jessie now...) , maybe this has changed?
If yes, using unique_ptr in lot of places would help a lot. Otherwise, using auto_ptr (which is deprecated, because of the arrival of unique_ptr, not because it's worse than new!) would be better than doing those malloc, realloc, and new stuff. 
I would like to know if you would accept to replace almost all occurrences of new by smart pointers. Either recent ones (unique_ptr) or old ones (deprecated in favor of recent ones, auto_ptr, ok, but still avoid to manually manage memory for nothing, especially when there are errors).

**Comment by authenticate** [comment]
Hi.

What were you doing in the game during the 3 hours where the memory usage increased to 3 gigs?  Were you actively playing the game, doing battles and quests?  Or was the game sitting and idling in some state for a period of time?  If possible, more information might help direct us to the problem.

I just checked the current build with heap and OpenGL tools designed to hunt memory leaks.  I ran the game through its various large scale systems; nothing appeared to be incorrect.

Now, this does not mean everything in the code base is correct and all is well.  There's always issues in nearly every code base.  It's part of the magic of software.  However, I am pretty confident we are not leaking memory in a way or on a scale to create 3 gigs worth of loss.  (I have been wrong before though. :D)

So, that brings us back to the original question of what is actually going on.

I personally would be interested in doing an analysis of Lua.  Lua keeps an internal record of heap objects created with it.  This makes it easier to use these objects in various scripts or in the C++ code itself because the lifetime of these objects is not really well defined.  It's essentially doing its own memory management.  When a Lua context is closed or, in our case, the end of the program, Lua cleans them all up.

I am not sure if we force Lua to clean up these objects periodically at certain points during the game, like loading a battle or loading a new map.  Someone with more experience with the code and Lua itself would have to comment on this.  This could actually be a complicated problem.  Certain things, like the positions and states of NPCs, probably should be retained throughout larger portions of the game.

I could image a scenario where someone plays the game for a long time and causes the game engine to create a lot of these objects.  This could theoretically cause a large amount of memory bloat.  In addition, since Lua properly cleans up these objects, there would not be any actual memory anomalies for debugging tools to report.

Anyways, this is just my casual analysis.  It could be totally wrong.  Take it for what you will. :)

**Comment by bmorel** [comment]
> What were you doing in the game during the 3 hours where the memory usage increased to 3 gigs? Were you actively playing the game, doing battles and quests? Or was the game sitting and idling in some state for a period of time? If possible, more information might help direct us to the problem.

I was playing actively. Doing quests and fights. I think the main leak come from battle, because the memory consumption of my computer increased a lot when I had battles, as said by xosview on my second screen. Not a very good measure anyway: it depends on my attention on both screens.

> I just checked the current build with heap and OpenGL tools designed to hunt memory leaks. I ran the game through its various large scale systems; nothing appeared to be incorrect.

It might be because I am using a different configuration, too. I have found several bugs in Debian because of that.
I am using the official clang 3.5 compiler (which is not in Debian wheezy, aka old-stable now), and as many stuff as possible from Debian old-stable. Other libs I needed where imported from Debian stable (Jessie) and compiled whith Wheezy's libs. 
Can you give me hints about the tools (name and version) you used to track lack of problems, so that I could try them and help more?

> Now, this does not mean everything in the code base is correct and all is well. There's always issues in nearly every code base.

Hehe... that's what's makes it fun to play with code.
Code is written by humans, but code's authors almost never master their own code. I have written enough code to know that.
Sometimes when I read and fix some pieces of code, I can be aggressive in my comments, but I know that I did worse, several years, or months, ago.
At least I am able to recognize that at the most important thing about code is that it does what it is asked to. Even if it written poorly in my *current* opinion. Writing easy code needs technique. Arts without techniques are just bunch of craps. My opinion.

> However, I am pretty confident we are not leaking memory in a way or on a scale to create 3 gigs worth of loss. (I have been wrong before though. :D)

I want to trust you.
But... I am a C++ coder myself, and I know one thing: my computer never lies.
It does not mean the problems I have spotted comes from VT's code, but, you can not blame me to first think they come from there.
Other targets are the dependencies, which are libSDL* and other well known libraries. I might have found a bug on those, too. Code is as humans: never perfect. Never. But, I am most inclined to think the problem comes from VT, considering that when I close it, the 3 gigs are freed. And other facts I can't prove, without a screencast ;)

> I personally would be interested in doing an analysis of Lua.

I am a decent C++ programmer. I am not a good one.
I am also a free software contributor. Still, not a good one :) I simply help when I can, on software which are small enough for my skills.
 If you can give me pointers to learn how I can improve my skills, and fix the problems I have noticed, I will be happy to learn those technics. Just pointers. I learned programming by myself since so many years, I won't be afraid to learn more now.

> the lifetime of these objects is not really well defined.

This sounds actually weird to me. Lifetime not defined? So, who owns what? And this question might, or not, lead to memory leaks.

> It's essentially doing its own memory management. When a Lua context is closed or, in our case, the end of the program, Lua cleans them all up.

But what happens when the software is never closed? As in my case, I would not have closed the software if there were not those memleaks. They almost led me to a general shutdown (but, since I have a second screen, with monitors, I noticed that the problem was because of RAM exhaustion, .and close few softwares, to be able to kill VT and relaunch it).
Before you ask, I hate garbage collectors, because you never know when they'll do their job, or if they'll do their job someday. I prefer RAII. But it's only implemented by languages that does not says that there *must* be a garbage collector, AFAIK. Be careful of the troll :)

> I am not sure if we force Lua to clean up these objects periodically at certain points during the game, like loading a battle or loading a new map. Someone with more experience with the code and Lua itself would have to comment on this. This could actually be a complicated problem. Certain things, like the positions and states of NPCs, probably should be retained throughout larger portions of the game.
> 
> I could image a scenario where someone plays the game for a long time and causes the game engine to create a lot of these objects. This could theoretically cause a large amount of memory bloat. In addition, since Lua properly cleans up these objects, there would not be any actual memory anomalies for debugging tools to report when the program shuts down.
> 
> Anyways, this is just my casual analysis. It could be totally wrong. Take it for what you will. :)

I will take it as an interesting point of view.
Because I am lost between the variety of my computer configurations, some have swap files or swap partitions, some does not, some runs windows-based OS, some runs Linux, and...  it's a mess! So, It is not stupid to think that the memleaks I have found comes from a bug in a specific version of a specific library.

But, I do think that the memleak could be because of lack of respect to GRASP[1] principles in src/engine/video. So, I'm trying to fix that, and when I'll have understood what that last bit of code I am facing(*)  is doing, I will submit some change.

[1]: http://en.wikipedia.org/wiki/Object-oriented_programming#SOLID_and_GRASP_guidelines

*: there are some loops which are almost identical but that I were not able to understand what they are doing, yesterday. I am retrying, night is good for coding.

**Comment by Bertram25** [comment]
Hey there,

I've been reading the whole conversation and there are a lot things here. I'll try to give my own humble opinion and help if I can. :)

As for the memleak, the battleMode and the use of lua and, let's not forget about it, luabind, here is what I know:
- The battleMode is using a pool named BattleMedia to get common media files without reloading them again and again. 
- Lua/luabind is indeed using a gc for own handled object instances. I've already work on this and already reduced a bit IMHO the gc overhead by forcing to renew the table namespace when opening a new mapmode for instance, permitting to tell the gc the instances in the previous tables can be dropped.
I've also added a function to force the gc to trigger. See: https://github.com/Bertram25/ValyriaTear/blob/master/src/engine/script/script.h#L304
There are a few other functions to know which tables are still open and such. Feel free to add a few debug calls here and there and correct me if you see something relevant. :)

- Note that there is also a gloal media pool used for common graphics and sounds. 

All that is the theory of course, and certain things may have gone wrong.

As for the use of unique_ptr, I must say I have nothing against c++11 at all, (since we're using it more and more broadly now) and the new features are good, I must say. It's just I didn't want to introduce the new coding standard, at least not until it's well supported on smaller and embedded platforms. :)

I would also rather avoid hiding problems with shared/unique_ptr unless ot's proven the best solution.

> But, I do think that the memleak could be because of lack of respect to GRASP1 principles in src/engine/video. So, I'm trying to fix that, and when I'll have understood what that last bit of code I am facing(*) is doing, I will submit some change.

Feel free to propose improvements as you see fit. We'll discuss them in the corresponding PR. :)

Best regards, :D



Best regards,

**Comment by bmorel** [comment]
I have searched a little.
Refactored the ImageMemory so that it now hides it's properties (I hate the idea of a class having a pointer publicly accessible. There is no longer C memory management calls into video.). For what I have learn, it might be related to the management system. There are a lot of calls to ImageMemory's constructor. Something like 3000+ in 10-15 minutes (by dying 3-4 times against andromelius. This is a nice place to hunt this bug, fast deaths possible.).
So, maybe issues with the stuff which manages images. Maybe not my memory problem, too. Still searching.

>  and the new features are good, I must say. 

I do like a lot the new for syntax. Lambdas can be quite useful too, they allow to use the algorithms much more easily than before. When using static arrays, the new container is nice too: standard algos on static arrays, that's useful.

> I would also rather avoid hiding problems with shared/unique_ptr unless ot's proven the best solution.

Shared_ptr can make problems because is wrongly used, it can lead to circular references. Java programmers know the kind of problems it can make hehe. But unique_ptr is just like a raw pointer, except it can not be copied (only moved, but it is indeed possible to access it's underlying pointer). It provides some nice features, like, automatic cleaning when it goes out of scope, and this cleaning can be customized. For example, you can say it to clean the SDL_Surface it hides with SDL_FreeSurface.

**Comment by Bertram25** [comment]
> Refactored the ImageMemory so that it now hides it's properties (I hate the idea of a class having a pointer publicly accessible. There is no longer C memory management calls into video.). 

This is a good thing. I've also spent a lot of time cleaning such bad practices in other classes. Feel free to go down that road. :)

> I do like a lot the new for syntax. Lambdas can be quite useful too, they allow to use the algorithms much more easily than before. When using static arrays, the new container is nice too: standard algos on static arrays, that's useful.

Yep, completely agreed. The thing is, I wish it could be more broadly supported. (And it should be the case soon enough. Then, we'll talk. ;})

> [edit] just found a better way to test the memory problem: simply change areas, it can be done rather quickly.

So, there is still a problem of memory management when changing maps. What happens if you force the gc to trigger at the end of the mapmode loading code, for instance?

**Comment by Bertram25** [comment]
Ah, btw, I'll have a look at your visible changes soon (promise). Hopefully, if they mimic the existing code guidelines enough and are tested a bit, we can merge them. :)
I have my own changes to merge btw. ;)

**Comment by bmorel** [comment]
I tried to follow guidelines as much as possible, but those changes are not entirely clean. They even contain comments that are simple here to remind me to remove them in a final commit before PR.
I tried my changes, but not in long gaming sessions. I won't do long sessions until I still think the RAM problem is gone.

**Comment by Bertram25** [comment]
k, good luck :) I'll be around if you need me.

**Comment by Bertram25** [comment]
For self-reference: 
https://github.com/bmorel/ValyriaTear/commit/b31a3e06a52f55cabd045bd0877147d3054af477

**Comment by Bertram25** [comment]
For reference, I consider #354 a duplicate of this one. As this one has got more discussion, I'll keep it alive while being younger.
Also, I've just moved the code to initial c++11 support and made luabind use unique_ptr instead of auto_ptr in 7fdc6ec1be56e982c65d0a622ea5bc021bd86cc2

Feel free to make a PR when you have code you want to share. :)

**Comment by bmorel** [comment]
I am still unable to determine the cause of that leak. I first thought it might be because of loading the same images multiple times and never freeing memory, but now I no longer think it is the cause. I guess I could clean what I did on my refactoring branch and push it anyway. I'll try to do that this week.

**Comment by Bertram25** [comment]
>  I guess I could clean what I did on my refactoring branch and push it anyway. I'll try to do that this week.

Yes, please do so. Determining the root cause of this specific issue can become easier when we avoid errors by design. :)

**Comment by Bertram25** [comment]
To make this sensitive issue advance, I guess the first step would be to create a function dumping the variables on the lua stack, so that we'll know more easily whether the lua stack is actually growing and what is there consuming memory uselessly.
It would also permit to know whether the GC triggering function is actually having an effet on that part.

@bmorel Do you still want to clean up your branch or may I import the changes and finish the job myself?

**Comment by Bertram25** [comment]
Superseded by #480 as importing @bmorel 's worthy changes is part of it.
